tests/regression/test_mesh_vertexonly.py

```diff
@@ -0,0 +1,333 @@
+
+
+# Mesh usage tests
+
+
+def functionspace_tests(vm, family, degree):
+    # Prep: Get number of cells
+    num_cells_mpi_global = MPI.COMM_WORLD.allreduce(vm.num_cells(), op=MPI.SUM)
+    # Can create function space
+    V = FunctionSpace(vm, family, degree)
+    # Can create function on function spaces
+    f = Function(V)
+    g = Function(V)
+    # Can interpolate and Galerkin project onto functions
+    gdim = vm.geometric_dimension()
+    if gdim == 1:
+        x, = SpatialCoordinate(vm)
+        f.interpolate(x)
+        g.project(x)
+    elif gdim == 2:
+        x, y = SpatialCoordinate(vm)
+        f.interpolate(x*y)
+        g.project(x*y)
+    elif gdim == 3:
+        x, y, z = SpatialCoordinate(vm)
+        f.interpolate(x*y*z)
+        g.project(x*y*z)
+    # Get exact values at coordinates with maintained ordering
+    assert np.shape(f.dat.data_ro)[0] == np.shape(vm.coordinates.dat.data_ro)[0]
+    assert np.allclose(f.dat.data_ro, np.prod(vm.coordinates.dat.data_ro, 1))
+    # Galerkin Projection of expression is the same as interpolation of
+    # that expression since both exactly point evaluate the expression.
+    assert np.allclose(f.dat.data_ro, g.dat.data_ro)
+    # Assembly works as expected - global assembly (integration) of a
+    # constant on a vertex only mesh is evaluation of that constant
+    # num_vertices (globally) times
+    f.interpolate(Constant(2))
+    assert np.isclose(assemble(f*dx), 2*num_cells_mpi_global)
+
+
+def vectorfunctionspace_tests(vm, family, degree):
+    # Prep: Get number of cells
+    num_cells_mpi_global = MPI.COMM_WORLD.allreduce(vm.num_cells(), op=MPI.SUM)
+    # Can create function space
+    V = VectorFunctionSpace(vm, family, degree)
+    # Can create functions on function spaces
+    f = Function(V)
+    g = Function(V)
+    # Can interpolate and Galerkin project onto functions
+    x = SpatialCoordinate(vm)
+    f.interpolate(2*as_vector(x))
+    g.project(2*as_vector(x))
+    # Get exact values at coordinates with maintained ordering
+    assert np.shape(f.dat.data_ro)[0] == np.shape(vm.coordinates.dat.data_ro)[0]
+    assert np.allclose(f.dat.data_ro, 2*vm.coordinates.dat.data_ro)
+    # Galerkin Projection of expression is the same as interpolation of
+    # that expression since both exactly point evaluate the expression.
+    assert np.allclose(f.dat.data_ro, g.dat.data_ro)
+    # Assembly works as expected - global assembly (integration) of a
+    # constant on a vertex only mesh is evaluation of that constant
+    # num_vertices (globally) times. Note that we get a vertex cell for
+    # each geometric dimension so we have to sum over geometric
+    # dimension too.
+    gdim = vm.geometric_dimension()
+    if gdim == 1:
+        f.interpolate(Constant((1,)))
+    if gdim == 2:
+        f.interpolate(Constant((1, 1)))
+    if gdim == 3:
+        f.interpolate(Constant((1, 1, 1)))
+    assert np.isclose(assemble(inner(f, f)*dx), num_cells_mpi_global*gdim)
```

Comment:
These tests all do a lot of things. Can you split them into pieces?

Reply:
It's not obvious to me how I might split them up given that `functionspace_tests` and `vectorfunctionspace_tests` each need the same setup for a given mesh, family and degree.

Reply:
That's what fixtures are for [you move the setup into reusable fixtures]. If you want to save the setup cost, you can say `@pytest.fixture(scope="module")` and now the fixture will only be consed once (rather than once per test).

The thing with having lots of these assertions in one go is that earlier failures hide later ones.

Reply:
Here I need to generate a variable number (depending on the spatial dimension) of return values for each fixture and I'm going need fixtures of fixtures. I understand the motivation for splitting these tests up but it just seems way too complicated to work out. If it fails one of these then that's already a show stopper...

Reply:
Assuming no further objections raised. Resolving.